api/app.go

```diff
@@ -10,56 +10,72 @@
 package api
 
 import (
+	"context"
 	"fmt"
+	"net"
+	"net/http"
 	"os"
 	"os/signal"
 	"strings"
+	"sync"
 	"syscall"
 	"time"
 
+	grpc_auth "github.com/grpc-ecosystem/go-grpc-middleware/auth"
+
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+
 	"github.com/topfreegames/podium/leaderboard"
 
 	"github.com/getsentry/raven-go"
-	"github.com/labstack/echo/engine"
-	"github.com/labstack/echo/engine/fasthttp"
-	"github.com/labstack/echo/engine/standard"
-	"github.com/labstack/echo/middleware"
 	newrelic "github.com/newrelic/go-agent"
 	"github.com/rcrowley/go-metrics"
 	"github.com/spf13/viper"
-	"github.com/topfreegames/extensions/echo"
-	extechomiddleware "github.com/topfreegames/extensions/echo/middleware"
 	"github.com/topfreegames/extensions/jaeger"
 	extnethttpmiddleware "github.com/topfreegames/extensions/middleware"
 	"github.com/topfreegames/podium/log"
 	"go.uber.org/zap"
+
+	api "github.com/topfreegames/podium/proto/podium/api/v1"
 )
 
 // JSON type
 type JSON map[string]interface{}
 
 // App is a struct that represents a podium Application
 type App struct {
-	Debug        bool
-	Fast         bool
-	Port         int
-	Host         string
-	ConfigPath   string
-	Errors       metrics.EWMA
-	App          *echo.Echo
-	Engine       engine.Server
-	Config       *viper.Viper
-	Logger       zap.Logger
-	Leaderboards *leaderboard.Client
-	NewRelic     newrelic.Application
-	DDStatsD     *extnethttpmiddleware.DogStatsD
+	Debug                bool
+	Fast                 bool
+	HTTPPort             int
+	GRPCPort             int
+	HTTPEndpoint         string
+	GRPCEndpoint         string
+	httpReady, grpcReady chan bool
+	Host                 string
+	ConfigPath           string
+	Errors               metrics.EWMA
+	grpcServer           *grpc.Server
+	httpServer           *http.Server
+	Config               *viper.Viper
+	Logger               zap.Logger
+	Leaderboards         *leaderboard.Client
+	NewRelic             newrelic.Application
+	DDStatsD             *extnethttpmiddleware.DogStatsD
 }
 
-// GetApp returns a new podium Application
-func GetApp(host string, port int, configPath string, debug, fast bool, logger zap.Logger) (*App, error) {
+// GetApp returns a new podium Application.
+// If httpPort is sent as zero, a random port will be selected (the same will happen for grpcPort)
+func GetApp(host string, httpPort, grpcPort int, configPath string, debug bool, logger zap.Logger) (*App, error) {
```

Comment:
Usually in Go we would call this `NewApp`. (Or just `New` if the package name was `app`.)

Reply:
I like New. Looks much better.